File: packages/python/elizaos/features/advanced_capabilities/clipboard/services/clipboard_service.py
"""File-based clipboard storage service.

Provides write, read, search, list, and delete operations on markdown files
stored in the clipboard directory.
"""

from __future__ import annotations

import builtins
import contextlib
import logging
import os
import re
from datetime import datetime
from pathlib import Path
from typing import TYPE_CHECKING

from ..types import (
    ClipboardConfig,
    ClipboardEntry,
    ClipboardReadOptions,
    ClipboardSearchOptions,
    ClipboardSearchResult,
    ClipboardWriteOptions,
)

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime

logger = logging.getLogger(__name__)

DEFAULT_CLIPBOARD_CONFIG = ClipboardConfig(
    base_path=os.path.join(os.path.expanduser("~"), ".eliza", "clipboard"),
    max_file_size=1024 * 1024,
    allowed_extensions=[".md", ".txt"],
)


def _resolve_clipboard_config(
    config: ClipboardConfig | None = None,
    runtime: IAgentRuntime | None = None,
) -> ClipboardConfig:
    """Resolve config from runtime settings, env vars, and explicit config."""
    base = ClipboardConfig(
        base_path=DEFAULT_CLIPBOARD_CONFIG.base_path,
        max_file_size=DEFAULT_CLIPBOARD_CONFIG.max_file_size,
        allowed_extensions=list(DEFAULT_CLIPBOARD_CONFIG.allowed_extensions),
    )

    # Check runtime settings
    if runtime:
        bp = runtime.get_setting("CLIPBOARD_BASE_PATH") if hasattr(runtime, "get_setting") else None
        if isinstance(bp, str) and bp.strip():
            base.base_path = bp.strip()
        mfs = (
            runtime.get_setting("CLIPBOARD_MAX_FILE_SIZE")
            if hasattr(runtime, "get_setting")
            else None
        )
        if isinstance(mfs, (str, int, float)) and not isinstance(mfs, bool):
            with contextlib.suppress(ValueError, TypeError):
                base.max_file_size = int(mfs)

    # Check env vars
    env_bp = os.environ.get("CLIPBOARD_BASE_PATH")
    if env_bp and env_bp.strip():
        base.base_path = env_bp.strip()
    env_mfs = os.environ.get("CLIPBOARD_MAX_FILE_SIZE")
    if env_mfs:
        with contextlib.suppress(ValueError, TypeError):
            base.max_file_size = int(env_mfs)

    # Override with explicit config
    if config:
        if config.base_path:
            base.base_path = config.base_path
        if config.max_file_size:
            base.max_file_size = config.max_file_size
        if config.allowed_extensions:
            base.allowed_extensions = config.allowed_extensions

    return base


class ClipboardService:
    """Service for managing file-based clipboard memories."""

    def __init__(
        self,
        runtime: IAgentRuntime,
        config: ClipboardConfig | None = None,
    ) -> None:
        self._config = _resolve_clipboard_config(config, runtime)

    async def _ensure_directory(self) -> None:
        Path(self._config.base_path).mkdir(parents=True, exist_ok=True)

    def _sanitize_filename(self, title: str) -> str:
        return re.sub(r"[^a-z0-9\s-]", "", title.lower()).replace(" ", "-").strip("-")[:100]

    def _get_file_path(self, entry_id: str) -> str:
        filename = entry_id if entry_id.endswith(".md") else f"{entry_id}.md"
        return os.path.join(self._config.base_path, filename)

    def _get_entry_id(self, filename: str) -> str:
        return os.path.splitext(os.path.basename(filename))[0]

    async def write(
        self,
        title: str,
        content: str,
        options: ClipboardWriteOptions | None = None,
    ) -> ClipboardEntry:
        """Write or append content to a clipboard entry."""
        await self._ensure_directory()
        options = options or ClipboardWriteOptions()

        entry_id = self._sanitize_filename(title)
        file_path = self._get_file_path(entry_id)
        now = datetime.now()

        created_at = now
        if os.path.exists(file_path) and options.append:
            existing = await self.read(entry_id)
            final_content = f"{existing.content}\n\n---\n\n{content}"
            created_at = existing.created_at
        else:
            tags_line = f"tags: [{', '.join(options.tags)}]" if options.tags else ""
            frontmatter_parts = [
                "---",
                f'title: "{title}"',
                f"created: {now.isoformat()}",
                f"modified: {now.isoformat()}",
            ]
            if tags_line:
                frontmatter_parts.append(tags_line)
            frontmatter_parts.extend(["---", ""])
            frontmatter = "\n".join(frontmatter_parts)
            final_content = f"{frontmatter}\n{content}"

        # Check file size
        if len(final_content.encode("utf-8")) > self._config.max_file_size:
            raise ValueError(
                f"Content exceeds maximum file size of {self._config.max_file_size} bytes"
            )

        with open(file_path, "w", encoding="utf-8") as f:
            f.write(final_content)

        logger.info("[ClipboardService] Wrote entry: %s", entry_id)

        return ClipboardEntry(
            id=entry_id,
            path=file_path,
            title=title,
            content=final_content,
            created_at=created_at,
            modified_at=now,
            tags=options.tags,
        )

    async def read(
        self,
        entry_id: str,
        options: ClipboardReadOptions | None = None,
    ) -> ClipboardEntry:
        """Read a clipboard entry by ID."""
        options = options or ClipboardReadOptions()
        file_path = self._get_file_path(entry_id)

        if not os.path.exists(file_path):
            raise FileNotFoundError(f"Clipboard entry not found: {entry_id}")

        stat = os.stat(file_path)
        with open(file_path, encoding="utf-8") as f:
            content = f.read()

        # Handle line range
        if options.from_line is not None or options.lines is not None:
            lines = content.split("\n")
            from_idx = max(0, (options.from_line or 1) - 1)
            num_lines = options.lines or (len(lines) - from_idx)
            content = "\n".join(lines[from_idx : from_idx + num_lines])

        # Parse frontmatter
        title = entry_id
        tags: list[str] = []
        created_at = datetime.fromtimestamp(stat.st_ctime)

        fm_match = re.match(r"^---\n([\s\S]*?)\n---", content)
        if fm_match:
            fm = fm_match.group(1)
            title_match = re.search(r'title:\s*"?([^"\n]+)"?', fm)
            tags_match = re.search(r"tags:\s*\[([^\]]+)\]", fm)
            created_match = re.search(r"created:\s*(.+)", fm)
            if title_match:
                title = title_match.group(1)
            if tags_match:
                tags = [t.strip() for t in tags_match.group(1).split(",")]
            if created_match:
                with contextlib.suppress(ValueError):
                    created_at = datetime.fromisoformat(created_match.group(1).strip())

        return ClipboardEntry(
            id=entry_id,
            path=file_path,
            title=title,
            content=content,
            created_at=created_at,
            modified_at=datetime.fromtimestamp(stat.st_mtime),
            tags=tags if tags else None,
        )

    async def exists(self, entry_id: str) -> bool:
        return os.path.exists(self._get_file_path(entry_id))

    async def list(self) -> list[ClipboardEntry]:
        """List all clipboard entries sorted by modification date."""
        await self._ensure_directory()
        entries: list[ClipboardEntry] = []

        try:
            for filename in os.listdir(self._config.base_path):
                ext = os.path.splitext(filename)[1]
                if ext not in self._config.allowed_extensions:
                    continue
                try:
                    entry_id = self._get_entry_id(filename)
                    entry = await self.read(entry_id)
                    entries.append(entry)
                except Exception as e:
                    logger.warning(
                        "[ClipboardService] Failed to read entry %s: %s",
                        filename,
                        e,
                    )
        except Exception as e:
            logger.error("[ClipboardService] Failed to list entries: %s", e)

        return sorted(entries, key=lambda e: e.modified_at, reverse=True)

    async def search(
        self,
        query: str,
        options: ClipboardSearchOptions | None = None,
    ) -> builtins.list[ClipboardSearchResult]:
        """Search clipboard entries using text matching."""
        options = options or ClipboardSearchOptions()
        entries = await self.list()
        results: list[ClipboardSearchResult] = []

        query_terms = [t for t in query.lower().split() if len(t) > 2]
        if not query_terms:
            return results

        for entry in entries:
            lines = entry.content.split("\n")
            content_lower = entry.content.lower()

            match_count = 0
            for term in query_terms:
                match_count += content_lower.count(term)

            if match_count == 0:
                continue

            score = min(1.0, match_count / (len(query_terms) * 3))
            if score < options.min_score:
                continue

            # Find best snippet
            best_start = 0
            best_end = min(len(lines), 5)
            for i, line in enumerate(lines):
                line_lower = line.lower()
                if any(term in line_lower for term in query_terms):
                    best_start = max(0, i - 2)
                    best_end = min(len(lines), i + 3)
                    break

            snippet = "\n".join(lines[best_start:best_end])

            results.append(
                ClipboardSearchResult(
                    path=entry.path,
                    start_line=best_start + 1,
                    end_line=best_end,
                    score=score,
                    snippet=snippet,
                    entry_id=entry.id,
                )
            )

        results.sort(key=lambda r: r.score, reverse=True)
        return results[: options.max_results]

    async def delete(self, entry_id: str) -> bool:
        """Delete a clipboard entry."""
        file_path = self._get_file_path(entry_id)
        try:
            os.unlink(file_path)
            logger.info("[ClipboardService] Deleted entry: %s", entry_id)
            return True
        except FileNotFoundError:
            return False

    async def get_summary(self) -> str:
        """Get a summary of all clipboard content."""
        entries = await self.list()
        if not entries:
            return "No clipboard entries found."

        parts = [f"**Clipboard Summary** ({len(entries)} entries)", ""]
        for entry in entries[:10]:
            preview = (
                re.sub(r"^---[\s\S]*?---\n*", "", entry.content, count=1)[:100]
                .replace("\n", " ")
                .strip()
            )
            parts.append(f"- **{entry.title}** ({entry.id})")
            parts.append(f"  {preview}{'...' if len(preview) >= 100 else ''}")
            parts.append(f"  _Modified: {entry.modified_at.strftime('%Y-%m-%d')}_")

        if len(entries) > 10:
            parts.append(f"\n_...and {len(entries) - 10} more entries_")

        return "\n".join(parts)

    def get_base_path(self) -> str:
        return self._config.base_path


def create_clipboard_service(
    runtime: IAgentRuntime,
    config: ClipboardConfig | None = None,
) -> ClipboardService:
    return ClipboardService(runtime, config)

File: packages/typescript/src/features/advanced-capabilities/clipboard/services/clipboardService.ts
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import type { IAgentRuntime } from "../../../../types/index.ts";
import { logger } from "../../../../types/index.ts";
import type {
	ClipboardConfig,
	ClipboardEntry,
	ClipboardReadOptions,
	ClipboardSearchOptions,
	ClipboardSearchResult,
	ClipboardWriteOptions,
} from "../types.ts";

export const DEFAULT_CLIPBOARD_CONFIG: ClipboardConfig = {
	basePath: path.join(os.homedir(), ".eliza", "clipboard"),
	maxFileSize: 1024 * 1024, // 1MB
	allowedExtensions: [".md", ".txt"],
};

function readStringSetting(
	runtime: IAgentRuntime | undefined,
	key: string,
): string | null {
	const direct = runtime?.getSetting?.(key);
	if (typeof direct === "string" && direct.trim()) {
		return direct.trim();
	}
	const envValue = process.env[key];
	return typeof envValue === "string" && envValue.trim()
		? envValue.trim()
		: null;
}

function readNumberSetting(
	runtime: IAgentRuntime | undefined,
	key: string,
): number | null {
	const direct = runtime?.getSetting?.(key);
	if (typeof direct === "number" && Number.isFinite(direct) && direct > 0) {
		return direct;
	}
	if (typeof direct === "string" && direct.trim()) {
		const parsed = Number(direct);
		if (Number.isFinite(parsed) && parsed > 0) {
			return parsed;
		}
	}
	const envValue = process.env[key];
	if (typeof envValue === "string" && envValue.trim()) {
		const parsed = Number(envValue);
		if (Number.isFinite(parsed) && parsed > 0) {
			return parsed;
		}
	}
	return null;
}

export function resolveClipboardConfig(
	config?: Partial<ClipboardConfig>,
	runtime?: IAgentRuntime,
): ClipboardConfig {
	const basePath = readStringSetting(runtime, "CLIPBOARD_BASE_PATH");
	const maxFileSize = readNumberSetting(runtime, "CLIPBOARD_MAX_FILE_SIZE");

	return {
		...DEFAULT_CLIPBOARD_CONFIG,
		...(basePath ? { basePath } : {}),
		...(maxFileSize ? { maxFileSize } : {}),
		...config,
	};
}

/**
 * Service for managing file-based clipboard memories.
 * Provides write, read, search, list, and delete operations.
 */
export class ClipboardService {
	private config: ClipboardConfig;

	constructor(runtime: IAgentRuntime, config?: Partial<ClipboardConfig>) {
		this.config = resolveClipboardConfig(config, runtime);
	}

	/**
	 * Ensures the clipboard directory exists
	 */
	private async ensureDirectory(): Promise<void> {
		try {
			await fs.mkdir(this.config.basePath, { recursive: true });
		} catch (error) {
			const errorMsg = error instanceof Error ? error.message : String(error);
			logger.error("[ClipboardService] Failed to create directory:", errorMsg);
			throw error;
		}
	}

	/**
	 * Generates a safe filename from a title
	 */
	private sanitizeFilename(title: string): string {
		return title
			.toLowerCase()
			.replace(/[^a-z0-9\s-]/g, "")
			.replace(/\s+/g, "-")
			.replace(/-+/g, "-")
			.substring(0, 100);
	}

	/**
	 * Gets the full path for a clipboard entry
	 */
	private getFilePath(id: string): string {
		const filename = id.endsWith(".md") ? id : `${id}.md`;
		return path.join(this.config.basePath, filename);
	}

	/**
	 * Extracts entry ID from a filename
	 */
	private getEntryId(filename: string): string {
		return path.basename(filename, path.extname(filename));
	}

	/**
	 * Writes or appends content to a clipboard entry
	 */
	async write(
		title: string,
		content: string,
		options: ClipboardWriteOptions = {},
	): Promise<ClipboardEntry> {
		await this.ensureDirectory();

		const id = this.sanitizeFilename(title);
		const filePath = this.getFilePath(id);
		const now = new Date();

		let finalContent: string;
		let createdAt = now;

		const exists = await this.exists(id);
		if (exists && options.append) {
			const existing = await this.read(id);
			finalContent = `${existing.content}\n\n---\n\n${content}`;
			createdAt = existing.createdAt;
		} else {
			// Create frontmatter with metadata
			const tagsLine = options.tags?.length
				? `tags: [${options.tags.join(", ")}]`
				: "";
			const frontmatter = [
				"---",
				`title: "${title}"`,
				`created: ${now.toISOString()}`,
				`modified: ${now.toISOString()}`,
				tagsLine,
				"---",
				"",
			]
				.filter(Boolean)
				.join("\n");

			finalContent = `${frontmatter}\n${content}`;
		}

		// Check file size
		if (
			Buffer.byteLength(finalContent, "utf8") >
			(this.config.maxFileSize ?? 1024 * 1024)
		) {
			throw new Error(
				`Content exceeds maximum file size of ${this.config.maxFileSize} bytes`,
			);
		}

		await fs.writeFile(filePath, finalContent, "utf8");

		logger.info(`[ClipboardService] Wrote entry: ${id}`);

		return {
			id,
			path: filePath,
			title,
			content: finalContent,
			createdAt,
			modifiedAt: now,
			tags: options.tags,
		};
	}

	/**
	 * Reads a clipboard entry by ID
	 */
	async read(
		id: string,
		options: ClipboardReadOptions = {},
	): Promise<ClipboardEntry> {
		const filePath = this.getFilePath(id);

		try {
			const stat = await fs.stat(filePath);
			let content = await fs.readFile(filePath, "utf8");

			// Handle line range reading
			if (options.from !== undefined || options.lines !== undefined) {
				const lines = content.split("\n");
				const fromLine = Math.max(1, options.from ?? 1) - 1; // Convert to 0-indexed
				const numLines = options.lines ?? lines.length - fromLine;
				content = lines.slice(fromLine, fromLine + numLines).join("\n");
			}

			// Parse frontmatter for metadata
			const frontmatterMatch = content.match(/^---\n([\s\S]*?)\n---/);
			let title = id;
			let tags: string[] = [];
			let createdAt = stat.birthtime;

			if (frontmatterMatch) {
				const frontmatter = frontmatterMatch[1];
				const titleMatch = frontmatter.match(/title:\s*"?([^"\n]+)"?/);
				const tagsMatch = frontmatter.match(/tags:\s*\[([^\]]+)\]/);
				const createdMatch = frontmatter.match(/created:\s*(.+)/);

				if (titleMatch) title = titleMatch[1];
				if (tagsMatch) tags = tagsMatch[1].split(",").map((t) => t.trim());
				if (createdMatch) createdAt = new Date(createdMatch[1]);
			}

			return {
				id,
				path: filePath,
				title,
				content,
				createdAt,
				modifiedAt: stat.mtime,
				tags,
			};
		} catch (error) {
			if ((error as NodeJS.ErrnoException).code === "ENOENT") {
				throw new Error(`Clipboard entry not found: ${id}`);
			}
			throw error;
		}
	}

	/**
	 * Checks if a clipboard entry exists
	 */
	async exists(id: string): Promise<boolean> {
		const filePath = this.getFilePath(id);
		try {
			await fs.access(filePath);
			return true;
		} catch {
			return false;
		}
	}

	/**
	 * Lists all clipboard entries
	 */
	async list(): Promise<ClipboardEntry[]> {
		await this.ensureDirectory();

		try {
			const files = await fs.readdir(this.config.basePath);
			const entries: ClipboardEntry[] = [];

			for (const file of files) {
				const ext = path.extname(file);
				if (!this.config.allowedExtensions?.includes(ext)) continue;

				try {
					const id = this.getEntryId(file);
					const entry = await this.read(id);
					entries.push(entry);
				} catch (error) {
					const errorMsg =
						error instanceof Error ? error.message : String(error);
					logger.warn(
						`[ClipboardService] Failed to read entry ${file}:`,
						errorMsg,
					);
				}
			}

			// Sort by modified date, most recent first
			return entries.sort(
				(a, b) => b.modifiedAt.getTime() - a.modifiedAt.getTime(),
			);
		} catch (error) {
			const errorMsg = error instanceof Error ? error.message : String(error);
			logger.error("[ClipboardService] Failed to list entries:", errorMsg);
			return [];
		}
	}

	/**
	 * Searches clipboard entries using text matching
	 */
	async search(
		query: string,
		options: ClipboardSearchOptions = {},
	): Promise<ClipboardSearchResult[]> {
		const entries = await this.list();
		const results: ClipboardSearchResult[] = [];

		const maxResults = options.maxResults ?? 10;
		const minScore = options.minScore ?? 0.1;

		// Tokenize and lowercase the query
		const queryTerms = query
			.toLowerCase()
			.split(/\s+/)
			.filter((t) => t.length > 2);

		for (const entry of entries) {
			const lines = entry.content.split("\n");
			const contentLower = entry.content.toLowerCase();

			// Calculate relevance score based on term frequency
			let matchCount = 0;
			for (const term of queryTerms) {
				const regex = new RegExp(term, "gi");
				const matches = contentLower.match(regex);
				if (matches) matchCount += matches.length;
			}

			if (matchCount === 0) continue;

			// Calculate score (simple TF-based scoring)
			const score = Math.min(1, matchCount / (queryTerms.length * 3));
			if (score < minScore) continue;

			// Find the best matching snippet
			let bestSnippetStart = 0;
			let bestSnippetEnd = Math.min(lines.length, 5);

			for (let i = 0; i < lines.length; i++) {
				const lineLower = lines[i].toLowerCase();
				for (const term of queryTerms) {
					if (lineLower.includes(term)) {
						bestSnippetStart = Math.max(0, i - 2);
						bestSnippetEnd = Math.min(lines.length, i + 3);
						break;
					}
				}
			}

			const snippet = lines.slice(bestSnippetStart, bestSnippetEnd).join("\n");

			results.push({
				path: entry.path,
				startLine: bestSnippetStart + 1,
				endLine: bestSnippetEnd,
				score,
				snippet,
				entryId: entry.id,
			});
		}

		// Sort by score descending and limit results
		return results.sort((a, b) => b.score - a.score).slice(0, maxResults);
	}

	/**
	 * Deletes a clipboard entry
	 */
	async delete(id: string): Promise<boolean> {
		const filePath = this.getFilePath(id);

		try {
			await fs.unlink(filePath);
			logger.info(`[ClipboardService] Deleted entry: ${id}`);
			return true;
		} catch (error) {
			if ((error as NodeJS.ErrnoException).code === "ENOENT") {
				return false;
			}
			throw error;
		}
	}

	/**
	 * Gets a summary of all clipboard content
	 */
	async getSummary(): Promise<string> {
		const entries = await this.list();

		if (entries.length === 0) {
			return "No clipboard entries found.";
		}

		const summaryParts = [
			`**Clipboard Summary** (${entries.length} entries)`,
			"",
		];

		for (const entry of entries.slice(0, 10)) {
			const preview = entry.content
				.replace(/^---[\s\S]*?---\n*/m, "") // Remove frontmatter
				.substring(0, 100)
				.replace(/\n/g, " ")
				.trim();

			summaryParts.push(`- **${entry.title}** (${entry.id})`);
			summaryParts.push(`  ${preview}${preview.length >= 100 ? "..." : ""}`);
			summaryParts.push(
				`  _Modified: ${entry.modifiedAt.toLocaleDateString()}_`,
			);
		}

		if (entries.length > 10) {
			summaryParts.push(`\n_...and ${entries.length - 10} more entries_`);
		}

		return summaryParts.join("\n");
	}

	/**
	 * Gets the base path for clipboard files
	 */
	getBasePath(): string {
		return this.config.basePath;
	}
}

/**
 * Factory function to create a ClipboardService instance
 */
export function createClipboardService(
	runtime: IAgentRuntime,
	config?: Partial<ClipboardConfig>,
): ClipboardService {
	return new ClipboardService(runtime, config);
}
